DOC-1398: Connect preview - data type conversions (#1959)
RedisLabs/redislabs-docs: content/connect/data-types/_index.md

---
Title: Data type conversion
linkTitle: Data type conversion
description: Describes how Redis Connect converts source data types to Redis data types.
headerrange: "2"
weight: 100
alwaysopen: false
categories: ["Connect"]
aliases: /connect/datatypes/
         /connect/datatypes.md
         /connect/data-types/
         /connect/data-types.md
---
Redis Connect converts source data to Redis data types.  The specific conversions depend on the original source and data type.  

This section describes data type conversions from different sources, including:

> - Common/ANSI SQL types:
> - MySQL specific types:
> - Oracle specific types:
> - PostgreSQL specific types:
> - Microsoft SQL Server specific data types:

Reviewer: ```suggestion
- Common/ANSI SQL types
- MySQL specific types
- Oracle specific types
- PostgreSQL specific types
- Microsoft SQL Server specific data types
```

Currently, Redis Connect stores incoming values as hashes.  As a result, all supported incoming values are converted to a string type.  The initial conversion is handled by the connector (Debezium).

Fields with unsupported data types do not appear in the target Redis database.

{{<note>}}This information describes features currently in preview.  Behavior may change before general availability.{{</note>}}

## Common/ANSI SQL types {#ansi-types}

> This sections describes data types common to many relational databases systems, including:

Reviewer: ```suggestion
This section describes data types common to many relational databases, including:
```

Author: Fixed

Example: When `binary.handling.mode` is set to `bytes`, the value is saved as a binary string and then transformed into the Redis data type.
    
To illustrate, the value `hello` is added to the table as a binary string (`0x68656C6C6F`).  This is stored in the Redis Connect instance as `"aGVsbG8AAAAAAAAAAAAAAAAAAAA="`, which is also the value passed to the Redis database.
    
### bitstring

Not supported

### blob

Target data type (hash): bytes string

The binary value is either base64-encoded or hex-encoded, based on the value of the `binary.handling.mode` property in the connector configuration.

Example: When `binary.handling.mode` is set to `bytes`, the binary image file loaded into the source table is converted by Debezium to bytes, which are stored in the target Redis database as a bytes string.

### boolean

Target data type (hash): string

Example:  True values are stored as `"1"` (one); false values are stored as `"0"` (zero).

### char

Target data type (hash): string, includes support for [UTF-8](https://en.wikipedia.org/wiki/UTF-8) and [Unicode](https://home.unicode.org/basic-info/faq/)

> Example:  PostgreSQL fixed length strings, such as `char(14)`, are space padded to full length.  For example, `"hello world"` is saved as `"hello world  "` (two spaces are added to the end of the strng).  To learn more, see [Character types](https://www.postgresql.org/docs/current/datatype-character.html).

Reviewer: ```suggestion
Example:  PostgreSQL fixed-length strings, such as `char(14)`, are space padded to full length.  For example, `"hello world"` is saved as `"hello world  "` (two spaces are added to the end of the string).  To learn more, see [Character types](https://www.postgresql.org/docs/current/datatype-character.html).
```

Example: real-FLOAT(63), double precision - FLOAT(126)

The `decimal.handling.mode` Debezium connector configuration setting affects this.  When set to `double`, an input value of `-3.402E+38` is saved to the target Redis database as  `"-340200000000000000000000000000000000000"`.

### long_raw {#o-long-raw}

Not supported

### nchar {#o-nchar}

Target data type (hash): string supporting Unicode characters  

Example: An Oracle input value of "Peace in Hebrew is שלום" is saved to the Redis database as `"Peace in Hebrew is \xd7\xa9\xd7\x9c\xd7\x95\xd7\x9d"`.  Trailing spaces are added to fill the field value.

### nclob {#o-nclob}

Not supported

### number(p, s) {#o-number}

Target data type (hash): string  

Example: `"10385274000.32"`

### nvarchar {#o-nvarchar}

Target data type (hash): string supporting Unicode characters  

Example: An Oracle input value of "Peace in Hebrew is שלום" is saved to the Redis database as `"Peace in Hebrew is \xd7\xa9\xd7\x9c\xd7\x95\xd7\x9d"`.

### raw {#o-raw}

Not supported

### rowid {#o-rowid}

Target data type (hash): string  

Example: `"AAAR1QAAOAAAACFAAA"`

### timestamp with tz {#o-timestamp}

Target data type (hash): string  

> Example: Debezium converts an an Oracle input value of `2021-12-30 14:23:46` to `"2021-12-30T14:23:46+02:00"`.  This is stored in the target Redis database as `"1611878400000"`, which represents the milliseconds since the the Epoch.

Reviewer: ```suggestion
Example: Debezium converts an Oracle input value of `2021-12-30 14:23:46` to `"2021-12-30T14:23:46+02:00"`.  This is stored in the target Redis database as `"1611878400000"`, which represents the milliseconds since the the Epoch.
```

Author: Fixed

Example: When the `decimal.handling.mode` Debezium configuration parameter is set to `precision`, an input value of `922337203685477.5807` is converted to a binary string value of `"f/////////8="`.  This is saved to the target Redis database as `"922337203685477.5807"`.

### nchar {#s-nchar}

Target data type (hash): fixed-size string that can contain Unicode characters.

### nvarchar {#s-nvarchar}

Target data type (hash): variable-size string that can contain Unicode characters.

### numeric {#s-numeric}

Target data type (hash): string, ranging from `-10^38+1` to `10^38`

Affected by two configuration settings.  When `time.precision.mode` is set to `connect` and `decimal.handling.mode` equals `precision`, Debezium converts an input value of `1.00E+33` to the binary string `"SztMqFqGw1MAAAAAAAAAAA=="`.  This is saved to the target Redis database as `"1000000000000000000000000000000000"`.

### rowversion {#s-rowversion}

Target data type (hash): string representing an automatically generated, unique binary number typically used as a version number for the original table row.

Example: `"0x00000000000007D0"`

### smalldatetime {#s-smalldatetime}

> Target data type (hash): string representing the miliseconds since the Epoch and does not include the timezone.  

Reviewer: ```suggestion
Target data type (hash): string representing the milliseconds since the Epoch and does not include the timezone.  
```

Author: Fixed.

Example: An input value of `2018-06-20 15:13:16` is saved to the target Redis database as `"1529507580000"`.  The number of seconds (`16`) is not retained.

### smallmoney {#s-smallmoney}

Target data type (hash): string, ranging from `-214,748.3648` to `214,748.3647`

> Example: When the `decimal.handling.mode` Deezium configuraion parameter is set to `string`,an input value of `-214748.3648` is saved to the target Redis database as  `"-214748.3648"`

Reviewer: ```suggestion
Example: When the `decimal.handling.mode` Debezium configuration parameter is set to `string`, an input value of `-214748.3648` is saved to the target Redis database as  `"-214748.3648"`
```

### Spatial geometry types {#s-geometry}

Not supported

### Spatial geography types {#s-geography}

Not supported

### table {#s-table}

Not supported

### text {#s-text}

> Target data type (hash): string containing variable length Unicode data

Reviewer: ```suggestion
Target data type (hash): string containing variable-length Unicode data
```

Author: Fixed
